### PR TITLE
new: [BREAKING] Add support for LKE Control Plane ACL

### PR DESCRIPTION
**Pull request by lgarber-akamai**
## 📝 Description

This pull request adds support for the upcoming LKE Control Plane ACL feature, including the following changes:

#### Updated Endpoints

* GET `/lke/clusters` - Updated to include ACL under `control_plane` field
* GET `/lke/clusters/{cluster_id}` - Updated to include ACL under `control_plane` field
* POST `/lke/clusters` - Updated to include ACL under `control_plane` field
* PUT `/lke/clusters/{cluster_id}` - Updated to include ACL under `control_plane` field

#### New Endpoints

* GET `/lke/clusters/{cluster_id}/control_plane_acl`
* PUT `/lke/clusters/{cluster_id}/control_plane_acl`
* DELETE `/lke/clusters/{cluster_id}/control_plane_acl`

#### New Event Action Constants

* `lke_control_plane_acl_create`
* `lke_control_plane_acl_update`
* `lke_control_plane_acl_delete`

## ✔️ How to Test

The following test steps assume you have pulled down this repo locally and have configured your Linode account according to the guidelines commented on `TPT-2854`.

### Integration Testing

```
make fixtures ARGS="-run TestLKECluster_withACL"
```

### Manual Testing

In a linodego sandbox environment (e.g. dx-devenv), run the following:

```go
package main

import (
	"context"
	"fmt"
	"github.com/sanity-io/litter"
	"log"
	"os"

	"github.com/linode/linodego"
)

func main() {
	ctx := context.Background()

	client := linodego.NewClient(nil)
	client.SetToken(os.Getenv("LINODE_TOKEN"))

	valueTrue := true

	// Create an LKE cluster with an allowed control plane IP
	cluster, err := client.CreateLKECluster(
		ctx,
		linodego.LKEClusterCreateOptions{
			Label:      "linodego-manual-test",
			Region:     "us-mia",
			K8sVersion: "1.29",
			NodePools: []linodego.LKENodePoolCreateOptions{
				{
					Type:  "g6-standard-1",
					Count: 1,
				},
			},
			ControlPlane: &linodego.LKEClusterControlPlaneOptions{
				ACL: &linodego.LKEClusterControlPlaneACLOptions{
					Enabled: &valueTrue,
					Addresses: &linodego.LKEClusterControlPlaneACLAddressesOptions{
						IPv4: &[]string{"10.0.0.1/32"},
					},
				},
			},
		},
	)
	if err != nil {
		log.Fatal(err)
	}

	// Manually retrieve the ACL since it isn't yet returned
	// from the LKE cluster response endpoint
	acl, err := client.GetLKEClusterControlPlaneACL(ctx, cluster.ID)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("Cluster ID:", cluster.ID)
	fmt.Println("Cluster ACL Configuration:", litter.Sdump(acl))

	acl, err = client.UpdateLKEClusterControlPlaneACL(
		ctx,
		cluster.ID,
		linodego.LKEClusterControlPlaneACLUpdateOptions{
			ACL: linodego.LKEClusterControlPlaneACLOptions{
				Addresses: &linodego.LKEClusterControlPlaneACLAddressesOptions{
					IPv4: &[]string{"10.0.0.5/32"},
				},
			},
		},
	)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("Updated Cluster ACL Configuration:", litter.Sdump(acl))
}
```

2. Ensure the output looks similar to the following:

```
Cluster ID: 173797
Cluster ACL Configuration: &linodego.LKEClusterControlPlaneACLResponse{
  ACL: linodego.LKEClusterControlPlaneACL{
    Enabled: &true,
    Addresses: &linodego.LKEClusterControlPlaneACLAddresses{
      IPv4: []string{
        "10.0.0.1/32",
      },
      IPv6: nil,
    },
  },
}
Updated Cluster ACL Configuration: &linodego.LKEClusterControlPlaneACLResponse{
  ACL: linodego.LKEClusterControlPlaneACL{
    Enabled: &true,
    Addresses: &linodego.LKEClusterControlPlaneACLAddresses{
      IPv4: []string{
        "10.0.0.5/32",
      },
      IPv6: nil,
    },
  },
}

### CODE REVIEW

File: lke_clusters_acl.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// allowed IP ranges for an LKE cluster's control plane.
+type LKEClusterControlPlaneACLAddresses struct {
+	IPv4 []string `json:"ipv4,omitempty"`
+	IPv6 []string `json:"ipv6,omitempty"`
+}
+
+// LKEClusterControlPlaneACL describes the ACL configuration
+// for an LKE cluster's control plane.
+type LKEClusterControlPlaneACL struct {
+	Enabled   *bool                               `json:"enabled,omitempty"`
```
Comment (lgarber-akamai) [inline lke_clusters_acl.go:15]: Making `Enabled` a `*bool` feels a bit wrong but is necessary to make sure users can disable ACL using the PUT endpoint.

File: lke_clusters_acl.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	Addresses *LKEClusterControlPlaneACLAddresses `json:"addresses,omitempty"`
+}
+
+// LKEClusterControlPlaneACLUpdateOptions represents the options
+// available when updating the ACL configuration of an LKE cluster's
+// control plane.
+type LKEClusterControlPlaneACLUpdateOptions struct {
+	ACL LKEClusterControlPlaneACL `json:"acl"`
```
Comment (lgarber-akamai) [inline lke_clusters_acl.go:23]: The structure of this request body deviates a bit from what you'd expect but I did confirm that this is what the API accepts

File: test/go.mod [hunk 3]
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/jarcoal/httpmock v1.3.1
 	github.com/linode/linodego v1.30.0
 	github.com/linode/linodego/k8s v0.0.0-00010101000000-000000000000
+	github.com/stretchr/testify v1.9.0
```
Comment (lgarber-akamai) [inline test/go.mod:9]: This was originally added in the parent/child project branch

File: test/integration/lke_clusters_acl_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+					},
+				}
+			},
+		},
+		"fixtures/TestLKECluster_withACL",
+	)
+	require.NoError(t, err)
+	defer teardown()
+
+	// TODO: Not currently populated in response, uncomment when available
```
Comment (lgarber-akamai) [inline test/integration/lke_clusters_acl_test.go:36]: We'll need to report this issue internally

File: lke_clusters.go [hunk 5]
```diff
@@ -35,20 +35,20 @@ type LKECluster struct {
...
-	Region       string                     `json:"region"`
-	K8sVersion   string                     `json:"k8s_version"`
-	Tags         []string                   `json:"tags,omitempty"`
-	ControlPlane *LKEClusterControlPlane    `json:"control_plane,omitempty"`
+	NodePools    []LKENodePoolCreateOptions     `json:"node_pools"`
+	Label        string                         `json:"label"`
+	Region       string                         `json:"region"`
+	K8sVersion   string                         `json:"k8s_version"`
+	Tags         []string                       `json:"tags,omitempty"`
+	ControlPlane *LKEClusterControlPlaneOptions `json:"control_plane,omitempty"`
```
Comment (lgarber-akamai) [inline lke_clusters.go:43]: Breaking change #1, TBD during cafe

**Comment by lgarber-akamai** [comment]
As a team we decided to go ahead with this breaking change 🙂

File: lke_clusters_control_plane.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+// for the Client.GetLKEClusterControlPlaneACL(...) method.
+type LKEClusterControlPlaneACLResponse struct {
+	ACL LKEClusterControlPlaneACL `json:"acl"`
+}
+
+// LKEClusterControlPlaneACLAddressesOptions are the options used to
+// specify the allowed IP ranges for an LKE cluster's control plane.
+type LKEClusterControlPlaneACLAddressesOptions struct {
+	IPv4 *[]string `json:"ipv4,omitempty"`
+	IPv6 *[]string `json:"ipv6,omitempty"`
```
Comment (lgarber-akamai) [inline lke_clusters_control_plane.go:34]: A pointer to a slice _might_ be consider an anti-pattern but in this case we need omitempty to only apply to nil values.

File: lke_clusters_control_plane.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+// LKEClusterControlPlaneACL describes the ACL configuration
+// for an LKE cluster's control plane.
+type LKEClusterControlPlaneACL struct {
+	Enabled   bool                                `json:"enabled"`
+	Addresses *LKEClusterControlPlaneACLAddresses `json:"addresses"`
+}
+
+// LKEClusterControlPlaneACLResponse represents the response structure
+// for the Client.GetLKEClusterControlPlaneACL(...) method.
+type LKEClusterControlPlaneACLResponse struct {
```
Comment (lgarber-akamai) [inline lke_clusters_control_plane.go:26]: I didn't quite know what to name this response struct. For reference, this response body is unique to `GetLKEClusterControlPlaneACL(...)` and `UpdateLKEClusterControlPlaneACL(...)`.

File: lke_clusters_control_plane.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+type LKEClusterControlPlaneACLAddressesOptions struct {
+	IPv4 *[]string `json:"ipv4,omitempty"`
+	IPv6 *[]string `json:"ipv6,omitempty"`
+}
+
+// LKEClusterControlPlaneACLOptions represents the options used when
+// configuring an LKE cluster's control plane ACL policy.
+type LKEClusterControlPlaneACLOptions struct {
+	Enabled   *bool                                      `json:"enabled,omitempty"`
+	Addresses *LKEClusterControlPlaneACLAddressesOptions `json:"addresses,omitempty"`
```
Comment (lgarber-akamai) [inline lke_clusters_control_plane.go:35]: I'm slightly concerned about the UX impact of making all update fields pointers but I don't think there's a better way to do it. 

Maybe we could introduce a helper like this to make it bit cleaner?

```go
func Optional[T any](value T) *T {
	return &value
}

// Usage: linodego.Optional("test-string")
```

File: lke_clusters_control_plane.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+// given cluster's control plane.
+func (c *Client) GetLKEClusterControlPlaneACL(ctx context.Context, clusterID int) (*LKEClusterControlPlaneACLResponse, error) {
+	return doGETRequest[LKEClusterControlPlaneACLResponse](
+		ctx,
+		c,
+		formatAPIPath("lke/clusters/%d/control_plane_acl", clusterID),
+	)
+}
+
+// UpdateLKEClusterControlPlaneACL gets the ACL configuration for the
```
Comment (yec-akamai) [inline lke_clusters_control_plane.go:68]: nit: 
```suggestion
// UpdateLKEClusterControlPlaneACL updates the ACL configuration for the
```

**Review by yec-akamai — APPROVED** [review]
Great work! Tests passed locally 🎉

File: lke_clusters_control_plane.go (hunk 8 above)
Comment (lgarber-akamai) [reply lke_clusters_control_plane.go:35]: @yec-akamai Sounds good! I'll track adding that helper as a separate ticket/PR just to keep the scope of this PR reasonable 🙂

**Review by ykim-1 — APPROVED** [review]
LGTM, Tests pass locally. Should we add the required customer tag to all the test account?

**Comment by lgarber-akamai** [comment]
> LGTM, Tests pass locally. Should we add the required customer tag to all the test account?

@ykim-1 Great idea, yeah we should definitely add the customer tag until this reaches GA

**Pull request merged**